src/Chainweb/Pact/PactService.hs

```diff
@@ -500,57 +500,58 @@ execNewBlock mpAccess parent miner = {- handle onTxFailure $ -} do
         logInfo $ "execNewBlock, about to get call processFork: "
                 <> " (parent height = " <> sshow pHeight <> ")"
                 <> " (parent hash = " <> sshow pHash <> ")"
 
-        gasLimit <- view psBlockGasLimit
+        initState <- (\g -> BlockFill g mempty 0) <$> view psBlockGasLimit
 
-        newTrans <- getBlockTxs gasLimit
+        newTrans <- getBlockTxs initState
 
         -- NEW BLOCK COINBASE: Reject bad coinbase, always use precompilation
         (Transactions pairs cb) <- execTransactions False miner newTrans
           (EnforceCoinbaseFailure True)
           (CoinbaseUsePrecompiled True)
           pdbenv
 
-        (BlockFilling _ successPairs failures _rks) <-
-          refill pdbenv =<< foldM splitResults (BlockFilling gasLimit mempty mempty mempty) pairs
+        (BlockFilling _ successPairs failures) <-
+          refill pdbenv =<< foldM splitResults (BlockFilling initState mempty mempty) pairs
 
         liftIO $ mpaBadlistTx mpAccess (V.map gasPurchaseFailureHash failures)
 
         let !pwo = toPayloadWithOutputs miner (Transactions successPairs cb)
         return $! Discard pwo
 
-    refill pdbenv unchanged@(BlockFilling gasLimit _ _ _)=
+    refill pdbenv unchanged@(BlockFilling bfState _ _)=
 
-      if gasLimit <= 0 then pure unchanged else do
+      if _bfGasLimit bfState <= 0 then pure unchanged else do
 
-        newTrans <- getBlockTxs gasLimit
+        newTrans <- getBlockTxs bfState
         if V.null newTrans then pure unchanged else do
 
           pairs <- execTransactionsOnly miner newTrans pdbenv
 
-          r@(BlockFilling gasRemaining newPairs newFails _newRks) <-
+          r@(BlockFilling newState newPairs newFails) <-
                 foldM splitResults unchanged pairs
 
           -- LOOP INVARIANT: gas must decrease OR only non-zero failures
-          if (gasRemaining < gasLimit) || (V.null newPairs && not (V.null newFails))
+          if (_bfGasLimit newState < _bfGasLimit bfState) || (V.null newPairs && not (V.null newFails))
               then refill pdbenv r
               else throwM $ MempoolFillFailure $ "Invariant failure: " <>
-                   sshow (gasRemaining,gasLimit,V.length newTrans
+                   sshow (bfState,newState,V.length newTrans
                          ,V.length newPairs,V.length newFails)
 
 
 
-    splitResults (BlockFilling g success fails rks) (t,r) = case r of
-      Right cr -> enforceUnique rks (requestKeyToTransactionHash $ P._crReqKey cr)
-        >>= return . BlockFilling (g - fromIntegral (P._crGas cr)) (V.snoc success (t,cr)) fails
-      Left f -> enforceUnique rks (gasPurchaseFailureHash f)
-        >>= return . BlockFilling g success (V.snoc fails f)
+    splitResults (BlockFilling (BlockFill g rks i) success fails) (t,r) = case r of
+      Right cr -> enforceUnique rks (requestKeyToTransactionHash $ P._crReqKey cr) >>= \rks' ->
+        return $ BlockFilling (BlockFill (g - fromIntegral (P._crGas cr)) rks' (succ i))
```

Comment:
Might be good to make note here that this is the heuristic we're using to decrement the block gas limit as we build blocks using the *actual gas used by the transaction*, as opposed to the specified `GasLimit` for the individual tx.